### ISSUE TITLE
new rule: no-switch-statement

### ISSUE THREAD

**Issue by RebeccaStevens** [jonaskello/eslint-plugin-ts-immutable#10]
Switch statements are just fancy if statement but even less functional.

It's probably safe to extend `no-if-statement` to check for this rather than creating a new rule. @jonaskello Would you agree?

**Comment by jonaskello** [comment]
Adding to an existing rule will keep the number of rules down which probably is a good thing. Perhaps we can rename `no-if-statement` to something like `no-conditional-statement` and have options to ignore `if` or `switch`.

It might be interesting to document alternatives to `switch`. I use `switch` a lot together with exhaustive type checking to emulate pattern matching. However it is still a statement which is non-functional. There are proposals like [do-expression](https://github.com/tc39/proposal-do-expressions) and [pattern-matching](https://github.com/tc39/proposal-pattern-matching) but only at stage 1 so they are probably years away :-/. Of course ternary operator could also be used with exhaustive type checking. Perhaps that is the only real option for keeping it strictly functional today. Wrapping the `switch` statements in a function and doing return in each `case` is what I usually do to emulate it being an expression.

**Comment by jonaskello** [comment]
Perhaps the `no-conditional-statement` could have an option to ignore constructs where every branch of the conditional ends with `return`. This is probably the closes we can get to making `if` or `switch` an expression in typescript? I'm thinking either wrapping in an IIFE or in a separate function should be allowed, and in both cases the branches all have to end with `return`. For example this should be (optionally) allowed:

```typescript
const y = "a";
const x = (() => {
   switch(y) {
     case "a":
       return 1;
     case "b":
       return 2;
     default:
       return 0;
}})();
```

**Comment by RebeccaStevens** [comment]
Closed by #23
